TCES-43 Implemented React State Management + React Routing + Authentication Connection between Frontend and Backend

kevinle623 opened the pull request:
# Description

1. Authentication Connection between Frontend and Backend + remebers auth on refresh
2. React State Management for (userData, auth detection, job updates, managed users for admin etc)
4. React Routing across all components currently available in the main branch + route guard for authed users AND admin 
users

# Gallery

https://github.com/uoftblueprint/tces/assets/74465053/7afed483-398f-4633-9455-83478f9a5095


https://github.com/uoftblueprint/tces/assets/74465053/ffe74cfa-d8a1-4919-b0ac-80231ecbfa80



https://github.com/uoftblueprint/tces/assets/74465053/409b062a-b7ee-4a7b-b223-fc44b7e47ca2


https://github.com/uoftblueprint/tces/assets/74465053/e81af126-8638-4ddd-b68f-49229bfb5126







Issue # 43

## Type of change

- [ x ] New feature (non-breaking change which adds functionality)

# Checklist

Please delete options that are not applicable.

- [ x ] My code follows the style guidelines of this project (required)
- [ x ] I have performed a self-review of my code (required)
- [ x ] I have linted and formatted my code (required)
- [ x ] New and existing unit tests pass locally with my changes (required)
- [ x ] I have commented my code, particularly in hard-to-understand areas
- [ x ] I have made corresponding changes to the documentation
- [ x ] My changes generate no new warnings

EclipseIsDead commented on package.json:
```diff
@@ -5,6 +5,11 @@
         "install:client": "cd client && npm i",
-        "install:server": "cd server && npm i"
+        "install:server": "cd server && npm i",
+        "test": "jest"
```
should probably only have jest in server level package.json

annie951977 reviewed (changes requested):
I agree with Siddarth, looks good otherwise

EclipseIsDead reviewed (approved):
lgtm

annie951977 reviewed (approved):
pog

kevinle623 committed:
set up auth routing between login and dashboard

kevinle623 committed:
Merge branch 'main' into kevinle6231/tces-43-urd-clean-up-temporary

kevinle623 committed:
Merge branch 'main' into kevinle6231/tces-43-urd-clean-up-temporary

kevinle623 committed:
more state setup and routing

kevinle623 committed:
Merge branch 'main' into kevinle6231/tces-43-urd-clean-up-temporary

kevinle623 committed:
fixed style error

kevinle623 committed:
finalized routing

kevinle623 committed:
some refactor

kevinle623 committed:
fixed admin auth logic

kevinle623 committed:
set up actions for managedusers modification

kevinle623 committed:
implemented edit funtionality with local state

kevinle623 committed:
implemented add new user functionality in local state

kevinle623 committed:
added remove functionality to state

kevinle623 committed:
added confirm dialog logic

kevinle623 committed:
integrated confirm edit logic on edit user page

kevinle623 committed:
added a path to admin

kevinle623 committed:
reset initial start

kevinle623 committed:
set basic auth with frontend to backend

kevinle623 committed:
fixed env files

kevinle623 committed:
added more admin logic

kevinle623 committed:
set up bypass auth

kevinle623 committed:
fixed lint errors

kevinle623 committed:
navigate backed

kevinle623 committed:
added git ignore

kevinle623 committed:
deleted package-lock.json

kevinle623 committed:
delete package-lock

kevinle623 committed:
added back pack-lock.json

jordanjanakievski reviewed (commented):
Definitely a lot of things happening in this PR. I looked through it a few times to try and make sure that I did not miss anything, but nothing really stood out, especially since a lot of it was just making sure the page renders the fetched data.

Thank you for cleaning up some of the file structuring too! 🥇 

I think the navbar is good to be added in. I will take another look once you add that, but otherwise awesome stuff!

kevinle623 commented:
> Definitely a lot of things happening in this PR. I looked through it a few times to try and make sure that I did not miss anything, but nothing really stood out, especially since a lot of it was just making sure the page renders the fetched data.
> 
> Thank you for cleaning up some of the file structuring too! 🥇
> 
> I think the navbar is good to be added in. I will take another look once you add that, but otherwise awesome stuff!

Hey Jordan, thanks so much for the review. I really appreciate it. I will get the navbar merged in asap, hoping for by tomorrow evening Monday. Will keep you updated. Cheers.

kevinle623 committed:
Merge branch 'main' into kevinle6231/tces-43-urd-clean-up-temporary

jordanjanakievski commented:
> > Definitely a lot of things happening in this PR. I looked through it a few times to try and make sure that I did not miss anything, but nothing really stood out, especially since a lot of it was just making sure the page renders the fetched data.
> > Thank you for cleaning up some of the file structuring too! 🥇
> > I think the navbar is good to be added in. I will take another look once you add that, but otherwise awesome stuff!
> 
> Hey Jordan, thanks so much for the review. I really appreciate it. I will get the navbar merged in asap, hoping for by tomorrow evening Monday. Will keep you updated. Cheers.

Thanks Kevin! We really appreciate it

kevinle623 committed:
connected navbar to dashboard and admin dashboard

kevinle623 committed:
merged navbar + signout functionality

kevinle623 committed:
fixed warnings :)

kevinle623 committed:
remembers authentication on login :)

kevinle623 committed:
removed console.log and set up bypass auth

kevinle623 committed:
fixed lint

kevinle623 committed:
tweaks to bypass auth mode

kevinle623 committed:
removed bypass auth

jordanjanakievski commented on client/src/pages/admin-dashboard/index.jsx:
```diff
@@ -1,7 +1,32 @@
+    displayName: `${user.firstName} ${user.lastName}`,
+  }));
+
+  const removeUser = (userID) => {
```
Do we have a commented-out API call to the backend to get all users? I know you have some mock users for temporary use while the Users backend is in limbo

kevinle623 replied:
Hi Jordan, we don't have an endpoint to get all the users yet, not one that I was able to find. That's why for now, I've just been using state to do the logic, but once we get / setup the end point we can easily link it up.

jordanjanakievski commented on client/src/components/edit-user-component/index.jsx:
```diff
@@ -9,46 +11,37 @@ import {
-    const userDataJSON = JSON.stringify(userData);
-
-    // Replace url with target route
-    fetch("http://localhost:8000/edit", {
```
If we remove this, will we have a spot where the user data will be accessed to pre-populate the fields for the editing of a user?

kevinle623 replied:
Hi Jordan, the code that I removed from this file was there before I touched it, we don't have an end point to edit a user with connect to the backend yet so there was compiling issues when we enter the page. So I left it out because it would just lead to an error when we click the submit button. When we have / setup the end points for edit user, I will be able to link everything up again.

jordanjanakievski reviewed (commented):
Looking good Kevin. Just a couple comments on some data access stuff

jordanjanakievski reviewed (dismissed):
Great stuff Kevin! Big plays for the team to get some functionality for the user. Thank you for setting this stuff up!

kevinle623 committed:
Merge branch 'main' into kevinle6231/tces-43-urd-clean-up-temporary

kevinle623 committed:
Merge branch 'main' into kevinle6231/tces-43-urd-clean-up-temporary

kevinle623 committed:
connected user end points to main app in admin dashboard

kevinle623 committed:
fixed admin user search

kevinle623 committed:
tweaked auth guard routing

DanielDervishi commented on client/src/utils/api.js:
```diff
@@ -0,0 +1,145 @@
+    });
+    return response;
+  } catch (error) {
+    throw error;
```
probably remove the catch and throw of the same error in each of the api methods (unless there's a good reason to keep it)

kevinle623 replied:
I have an Error Screen Component set up, so I catch and throw the error so that the component that called this api function can catch the error, and the set the error.message screen to display to user and for us developers. For example, ` if (errorDisplay) return <ErrorComponent message={errorDisplay} />;` is an example where I have errorDisplay as a local state of the component and once I call my api function then I can catch the error and then display it to the user through this if statement.

DanielDervishi commented on client/src/App.jsx:
```diff
@@ -1,15 +1,168 @@
+              <Navigate to="/dashboard" />
+            ) : (
+              <Navigate to="/signin" />
+            )
```
Is there a reason we don't use the auth guard here?

kevinle623 replied:
Yeah you're right ill add this here.

DanielDervishi commented on client/src/App.jsx:
```diff
@@ -1,15 +1,168 @@
+            <AuthGuard
+              isAuthenticated={isAuthenticated}
+              loginUser={loginUser}
+              redirectUrl={adminRedirect}
```
Is there a reason for setting a redirectUrl here?

kevinle623 replied:
Yeah, so if we are on create-user or edit-user i was thinking if someone refresh or hard url navigates to this url, it would be best to navigate gate them to the admin dashboard instead of letting them go straight to edit user and create user, it's an extra param I added to this component in-case we want this type of behaviour. I think it makes sense instead of someone hard url navigating to something like "/admin/edit-user/49" blindly. It's better to only get to edit-user knowing that you clicked that specific  user from the admin dashboard to begin with. That's my thought process but let me know if I should remove.

DanielDervishi replied:
Ok that sounds good, we can leave it as it is.

DanielDervishi commented on client/src/App.jsx:
```diff
@@ -1,15 +1,168 @@
+              isPermitted={isAuthenticated}
+              redirect={notAuthRedirect}
+              redirectUrl={adminRedirect}
+            >
```
Maybe auth guard would be a better choice here?

DanielDervishi commented on client/src/components/navbar-component/Navbar.jsx:
```diff
@@ -1,9 +1,10 @@
 import NavbarButton from "./NavbarButton";
 
-function Navbar() {
+function Navbar({ isAdmin }) {
```
https://www.youtube.com/watch?v=B0vZGTGgIdc: Consider using an outlet within this component to prevent us from needing to import the navbar into each component that needs it in the future

DanielDervishi commented on client/src/components/create-user-component/index.jsx:
```diff
@@ -107,16 +98,20 @@ function CreateComponent() {
-            Submit
+            {isLoading ? "Submitting..." : "Submit"}
           </Button>
+          <ErrorMessage>{errorMessage}</ErrorMessage>
```
Due to the placement of the error message, this is what I see. I believe this is also the case with the edit component. <img width="1006" alt="image" src="https://github.com/uoftblueprint/tces/assets/58835213/403d031b-f607-4ab9-9321-83e88dc8c160">

kevinle623 replied:
Nice catch, I have moved it down here
<img width="1480" alt="image" src="https://github.com/uoftblueprint/tces/assets/74465053/241853cd-e95f-4d54-b070-c9cb0d53f1ee">

DanielDervishi commented on server/index.js:
```diff
@@ -1,8 +1,14 @@
 const port = 8000;
 
+const corsOption = {
+  origin: "http://localhost:3000",
```
Use FRONTEND_URL from the .env file in the backend here

kevinle623 replied:
Hi Daniel, the corsOption is only for localhost:3000 specifically so that we can communicate with the backend from our local frontend.

DanielDervishi commented on server/index.js:
```diff
@@ -19,25 +25,30 @@ const authRouter = require("./src/routes/auth");
-app.listen(port, () => {
-  console.log(`Example app listening on port ${port}`);
+  res.send("Helloooo World! Hi :)");
 });
```
This was sort of a dummy request I set up when creating the repo, it can be deleted 😄

DanielDervishi commented on server/src/routes/auth.js:
```diff
@@ -16,14 +15,36 @@ const isLoggedInRequestHandler = require("../controllers/auth/isLoggedIn");
+          email: user.email,
+          isAdmin: user.is_admin,
+        },
+      });
```
I think the authenticate method already invokes req.login: https://www.passportjs.org/concepts/authentication/login/

kevinle623 replied:
I set up my own response body on login so I can provide the frond end with info (id, firstname, last name, isadmin etc) to populate the currUser state.  Otherwise, by default it login will be invoked but would not have provided the required information for us in the frontend (isAdmin, userID etc). Please let me know if this is fine.

DanielDervishi commented on client/src/App.jsx:
```diff
@@ -1,15 +1,168 @@
+
+  // states defined at the very root of the react tree (will be passed down to contributing child components)
+  // User State
+  const [currUser, setCurrUser] = useState(mockUser);
```
Not sure but I don't think using a mock user is necessary anymore?

DanielDervishi commented on client/src/App.jsx:
```diff
@@ -1,15 +1,168 @@
+  const [jobUpdates] = useState(mockJobUpdates);
+
+  // Admin State
+  const [managedUsers, setManagedUsers] = useState(mockManagedUsers);
```
Same Idea as the mock user, shouldn't be necessary anymore

DanielDervishi commented on client/src/mock-data/mockManagedUsers.jsx:
```diff
@@ -0,0 +1,46 @@
+const mockManagedUsers = [
```
remove this file

DanielDervishi commented on client/src/mock-data/mockUser.jsx:
```diff
@@ -0,0 +1,9 @@
+const mockUser = {
```
remove this file

DanielDervishi commented on client/src/pages/create-user/index.jsx:
```diff
@@ -0,0 +1,28 @@
+    });
+  };
+  return <CreateComponent onAddUser={addUser} />;
+}
```
I think you can remove these methods as well as the setManagedUsers state given navigating to the admin dashboard after creating a user will trigger a reload of managedUsers anyways

DanielDervishi commented on client/src/pages/dashboard/index.styles.jsx:
```diff
@@ -0,0 +1 @@
+export {}
```
no other file references this afaik, we can remove it

DanielDervishi reviewed (commented):
Amazing work Kevin! All the code is well written and organized. This will serve as a great foundation for future prs!

kevinle623 committed:
cleaned up folder structure

kevinle623 committed:
used outlet

kevinle623 committed:
fixed circular progress animation

kevinle623 committed:
Merge branch 'main' into kevinle6231/tces-43-urd-clean-up-temporary

kevinle623 committed:
some more minor tweaks

kevinle623 committed:
Merge branch 'main' into kevinle6231/tces-43-urd-clean-up-temporary

kevinle623 committed:
fixed outlet

kevinle623 committed:
Merge branch 'main' into kevinle6231/tces-43-urd-clean-up-temporary

kevinle623 committed:
cleaned up auth

kevinle623 committed:
removed try catch in api calls and reverted back auth

kevinle623 committed:
fixed lint

kevinle623 committed:
fixed styling and left space below for error message

jordanjanakievski commented on client/.eslintrc.json:
```diff
@@ -14,6 +14,6 @@
         "react/react-in-jsx-scope": "off",
         "react/jsx-uses-react": "off",
-        "prettier/prettier": "error"
+        "prettier/prettier": "warn"
```
I think need to swap this back to error

kevinle623 replied:
Hi Jordan, Daniel and I agreed to keep this as warn about 2 months ago. We agreed that having "error" on will make it a hassle for developers to test / debug code because it won't let you compile unless you clean up the code. Either way, the github action I set up will raise a flag / failed run if there are any lint issues so I think it should be fine to keep it as "warn" for developers. If you are ok with this please let me know, if not please let me know too so I can swap it out asap. Thanks.

jordanjanakievski replied:
Ok. As long as the Actions blocks it I am good with that!

jordanjanakievski commented on client/src/components/edit-user-component/index.jsx:
```diff
@@ -9,46 +11,56 @@ import {
+    setConfirmEditDialog(false);
   };
 
   return (
```
Can we prepopulate the fields so that it is easier to make the edit to the user now that we have the users GET request added

jordanjanakievski reviewed (commented):
Everything is looking and working great. I think it's just a couple of small comments then good from my end.

- You can remove the db sessions file from the `var` directory in `server`. We have added db sessions to gitignore so just needs to be deleted from this branch when it gets merged in. Just like this previous branch. https://github.com/uoftblueprint/tces/tree/a5e4b18cf32d0503699de2c82164400de1f36af1/server/var/db

Thank you so much Kevin! It is really great to have some nice working functionality

kevinle623 committed:
pre-populated edit user form except for pass word

kevinle623 committed:
deleted session.db

kevinle623 commented:
> Everything is looking and working great. I think it's just a couple of small comments then good from my end.
> 
> * You can remove the db sessions file from the `var` directory in `server`. We have added db sessions to gitignore so just needs to be deleted from this branch when it gets merged in. Just like this previous branch. https://github.com/uoftblueprint/tces/tree/a5e4b18cf32d0503699de2c82164400de1f36af1/server/var/db
> 
> Thank you so much Kevin! It is really great to have some nice working functionality

Hey Jordan, thanks for the review. I have deleted the sessions.db file, I have also prepolated the edit user form, and I have addressed your request to convert prettier back to "error" from "warn". If there is anything else, please let me know.

DanielDervishi commented on client/src/components/admin-dashboard-component/index.jsx:
```diff
@@ -0,0 +1,278 @@
+
+  const handleConfirmDelete = async () => {
+    if (userToDelete !== null) {
+      setIsLoading(true);
```
Remove anything relating to the loading state in this component. This makes it so that whenever a user is deleted a loading page pops up. Removing this makes the deletion flow much smoother.

kevinle623 replied:
I agree thanks.

DanielDervishi commented on client/src/components/edit-user-component/index.jsx:
```diff
@@ -109,16 +126,32 @@ function EditComponent() {
           </CardContent>
```
For some reason I can't highlight the textfield component above this, but remove the requirement for password to be specified. Users may not want to edit the password of an existing user -- allow an empty value to indicate no password change. I believe the backend is already built out to handle the case where the password is left empty.

kevinle623 replied:
Sounds good, I removed the requirement for edit user.

DanielDervishi reviewed (commented):
Great work Kevin! After addressing these comments (and any that Jordan left on his last review) we should be good to get this pr merged in.

kevinle623 committed:
removed password requirement for edit user + removed loading logic in delete user in admin dashboard

kevinle623 commented:
> Great work Kevin! After addressing these comments (and any that Jordan left on his last review) we should be good to get this pr merged in.

Hey Daniel, thank you so much for the review. Nice catches, I have implemented the suggestions. :) If there is anything else please let me know.

DanielDervishi reviewed (approved):
Great Work Kevin!

jordanjanakievski reviewed (approved):
Thank you so much!
